text/0080-datetime-extension.md

```diff
@@ -0,0 +1,367 @@
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+### Alt. B: Pass results of time checks in the context
+
+Another workaround we have suggested is to simply handle date/time logic _outside_ of Cedar, and pass the results of checks in the `context`. For example, you could pass in fields like  `context.isWorkingHours` or `context.dayOfTheWeek`.
+
+Here are the previous examples rewritten to use additional context.
+
+**Only allow experienced, tenured persons from the Hardware Engineering department to see prototypes.**
+
+```cedar
+permit(
+  principal is User,
+  action == Action::"view",
+  resource in Folder::"device_prototypes"
+)
+when {
+  principal.department == "HardwareEngineering" &&
+  principal.jobLevel >= 10 &&
+  principal.hireDate <= context.minimumHiringDateForAccess
+};
+```
+
+Note: assumes `context.hireDate` and `context.minimumHiringDateForAccess` are `long` values (e.g., Unix time).
+
+**Only allow user "alice" to view JPEG photos for one week after creation.**
+
+```cedar
+permit(
+  principal == User::"alice",
+  action == PhotoOp::"view",
+  resource is Photo
+) when {
+  resource.fileType == "JPEG" &&
+  resource.creationDate >= context.oldestViewableDate
+};
+```
+
+Note: assumes `resource.creationDate` and `context.oldestViewableDate` are `long` values (e.g., Unix time).
+
+**Allow access from a certain IP address only between 9am and 6pm UTC.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.isWorkingHours
+};
+```
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  context.isTheWeekend
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  context.isMyBirthday && context.isLeapDay
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.afterBrexit &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Unresolved questions
+
+### Should we consider local time at all?
+
+Writing programs that involve dates and time is tricky. Standardizing on UTC representations for all dates and times trades achieving evaluator correctness more quickly at the expense of decreased readability for policy authors and auditors.
+
+### Should we provide a `currentTime` function?
+
+The examples above expect `currentTime` (or `now`) to be passed in the context during the authorization request. An alternative would be to provide a `currentTime()` extension function that computes the current time at policy evaluation time. This has a drawback, when testing policies, in that the evaluator must be somehow extended to mock the time.
```

Comment:
Worth noting here that we cannot provide `currentTime` at all.  That's because `currentTime` is stateful; it's not a pure function.  That is, `currentTime() == currentTime()` could return false, and in fact `currentTime()` is likely to return different values during the same authorization call over multiple policies that use it.

Reply:
Yes, exactly! currentTime() would also have to consider leap seconds, which we should definitely punt on. :)

I can just drop this section if it's easier.

Reply:
(I'll add your comment about purity, and maybe move to out of scope, instead)